client/rest/src/main/java/org/elasticsearch/client/Cancellable.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.client;
+
+import org.apache.http.client.methods.HttpRequestBase;
+
+import java.util.concurrent.CancellationException;
+
+/**
+ * Represents an operation that can be cancelled.
+ * Returned when executing async requests through {@link RestClient#performRequestAsync(Request, ResponseListener)}, so that the request
+ * can be cancelled if needed.
+ */
+public class Cancellable {
+
+    static final Cancellable NO_OP = new Cancellable(null) {
+        @Override
+        public synchronized void cancel() {
+        }
+
+        @Override
+        synchronized void runIfNotCancelled(Runnable runnable) {
+            throw new UnsupportedOperationException();
+        }
+    };
+
+    static Cancellable fromRequest(HttpRequestBase httpRequest) {
+        return new Cancellable(httpRequest);
+    }
+
+    private final HttpRequestBase httpRequest;
+
+    private Cancellable(HttpRequestBase httpRequest) {
+        this.httpRequest = httpRequest;
+    }
+
+    /**
+     * Cancels the on-going request that is associated with the current instance of {@link Cancellable}.
+     */
+    public synchronized void cancel() {
+        this.httpRequest.abort();
```

Comment:
maybe check here if the request is cancelled? Also does this need to be synchronized?

Reply:
I don't think I need to check if it's already cancelled. That check is already performed by the client, see `AbstractExecutionAwareRequest#abort`, hence the abort will do something only the first time it's called (unless the request gets reset).

It is synchronized to make sure that a request does not get cancelled when we are in between one attempt and the following one, in which case, if we have already reset the request but we have not yet called execute, the abort will do nothing as the cancellable ref is not set internally.

Reply:
True, probably not much that can go wrong here. Worst case you probably get some kind of exception from apache httpclient. Thanks for picking this up.

Reply:
the problem is that as a user you would call abort, which would do nothing and you don't get feedback on that. And any subsequent abort call would still do nothing as the request is marked aborted although calling abort did nothing. With the synchronized blocks we make sure that abort is only called when a cancellable ref is set internally, hence abort does something and is reliable Thank you for opening the issue and having a look at this PR.

Reply:
Clear. In any case good enough for my use case, which would be having this called from a [suspendCancellableCoRoutine](https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/suspend-cancellable-coroutine.html). This will work beautifully for that.

Reply:
I prefer Cancellable over Abortable, and the package will differ so it should be ok.

Reply:
`Abortable` is a terrible name indeed :-) .  
`Cancellable` sounds more like an interface, what about `CancellationHandler` ?

Reply:
We don't really need a new interface, a `Runnable` would be enough since the user can only trigger one action (cancel/abort). Something like:
````
Runnable cancel = client. performRequestAsync(...)
...
cancel.run();
````
The javadocs need to explain what cancellation means anyway so I don't think that a new interface adds anything.

Reply:
I am slightly in favour of having a specific class name, just because it's a publicly exposed API and Runnable is a bit too generic. I would not add a new interface (keep what we have as a class), as it would have a single impl and that is redundant. I don't have a strong opinion though. Curious to hear from others.